[FIX] 각종 버그 수정

jd99iam opened the pull request:
## 186

### 💡 PR 타입(하나 이상의 PR 타입을 선택해주세요)
* [ ] 기능 추가
* [ ] 기능 삭제
* [ ] 버그 수정
* [ ] 테스트코드 작성
* [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트

### 📲 반영 브랜치
ex) fix/186-bugs -> develop

### ❔ 변경 사항
**댓글 생성시 추가한 검증기능**
- 부모 댓글이 실제로 존재하는지
- 부모 댓글의 풀이랑 자식 댓글의 풀이랑 같은지
**리뷰 생성시 추가한 검증기능**
- 부모 리뷰가 실제로 존재하는지
- 부모 리뷰의 풀이랑 자식 댓글의 풀이랑 같은지 
- 부모 리뷰의 줄번호랑 자식 리뷰의 줄번호랑 같은지
- 리뷰의 줄번호가 1 <= 줄번호 <= 끝줄번호 인지

### 🔍 테스트 결과
추가된 검증 기능에 대한 테스트는 추후에 추가 하겠습니다

### 📰 참고 Reference
 
### ⚒️ 기타 사항

jd99iam committed:
feat: existById 추가

- 존재 여부만 확인하면 되므로 existById 추가

related to: #186

jd99iam committed:
feat: 예외 추가

추가하려는 대상의 부모가 없을 때 던질 예외 추가

related to: #186

jd99iam committed:
feat: 에러코드 추가

추가하려는 대상의 부모가 없을 때 던질 예외의 에러코드 추가

related to: #186

jd99iam committed:
fix: parentId 검증 추가

댓글을 작성할 때 parentId로 넘어오는 값에 대한 comment가 없으면 예외 던지도록 수정

related to: #186

jd99iam committed:
feat: 예외 추가

댓글 생성할 때 자식과 부모의 댓글의 다르면 던질 예외 추가

related to: #186

jd99iam committed:
feat: 예외 추가

댓글 생성할 때 자식과 부모의 댓글의 다르면 던질 예외의 에러코드 추가
related to: #186

jd99iam committed:
fix: 검증 로직 추가

댓글을 생성하려 할 때 생성하려는 댓글과 부모 댓글의 풀이가 같은지 검증하는 로직 추가

related to: #186

jd99iam committed:
feat: 리포지토리 existsById 추가

풀이가 존재하는지 확인하기 위한 메서드 추가

related to: #186

jd99iam committed:
fix: 리뷰 생성 시 검증 로직 추가

리뷰 생성에 추가로 필요한 검증 로직 추가
- 입력한 부모 id인 리뷰가 실제로 있는지 검증
- 입력한 부모 id인 리뷰의 풀이랑 지금 풀이랑 같은지 검증

related to: #186

jd99iam committed:
fix: 검증 로직에 부정 추가

헷갈려서 잘못 작성한 로직 수정
- 존재하지 않으면 예외를 던지는 것으로 수정

related to: #186

jd99iam committed:
fix: 던지는 예외 종류 수정

related to: #186

jd99iam committed:
feat: 리뷰 작성 검증 추가

리뷰 작성시 줄 번호 검증 기능 추가
- 에러코드 추가
- 예외 추가
- 줄 번호가 전체 코드의 줄 번호 안에 해당되는지 검증 메서드 추가

jd99iam committed:
fix: 코드 줄번호 검증 기능 수정

줄번호는 1부터 세니까 0이하일경우 예외 던지기로 수정

jd99iam committed:
feat: 줄 번호 검증 기능 추가

부모 리뷰가 있는 경우 부모 리뷰의 줄 번호와 새로 생성하러는 리뷰의 줄 번호가 같은지 검증
- 틀릴 경우 예외 던짐
- 예외 추가
- 에러코드 추가

jd99iam committed:
fix: 잘못 작성한 검증 로직 수정

부모 id가 아닌 풀이 id랑 비교하는 것으로 수정

jd99iam committed:
feat; 검증 로직 추가

내 풀이 상세보기에서 검증 로직 추가
- 서비스단에서 내 풀이인지 확인하는 로직 추가

related to: #186

jd99iam committed:
refactor: 미사용 import문 삭제

related to: #186

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/ReviewService.java:
```diff
@@ -67,7 +67,7 @@ private void validCodeLineNumber(WriteReviewRequest writeReviewRequest, int leng
     private void validSameSolution(WriteReviewRequest writeReviewRequest, Long solutionId) {
         if (writeReviewRequest.getParentId()==null) return;
-        if (solutionId!=reviewRepository.findById(writeReviewRequest.getParentId()).get().getParentId()){
+        if (solutionId!=reviewRepository.findById(writeReviewRequest.getParentId()).get().getSolution().getId()){
```
라인 하나에 반환값 너무 많은 것 같습니다. 
그리고 findByid를 하고 바로 get을 하고 있네요 만약에 null이면 NPE가 터지지 않을까요 ?

jd99iam replied:
> 라인 하나에 반환값 너무 많은 것 같습니다. 그리고 findByid를 하고 바로 get을 하고 있네요 만약에 null이면 NPE가 터지지 않을까요 ?

라인을 나눠서 가독성을 높이겠습니다
부모가 존재하는지 validParentExists에서 검증하고 나서 validSameSolution을 해서 NPE는 터지지 않습니다!

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/CommentService.java:
```diff
@@ -43,7 +43,8 @@ private Comment findById(Long id) {
     public ShowMyCommentsResponse showMyComments(Long memberId, Integer pageNumber) {
         validPageNumber(pageNumber);
-        Slice<Comment> commentPage = commentRepository.findMyPageByMemberId(PageRequest.of(pageNumber - 1, MY_COMMENTS_PAGE_SIZE), memberId);
+        Slice<Comment> commentPage = commentRepository.findMyPageByMemberId(
```
줄바꿈을 왜 한건가요 ?

jd99iam replied:
> 줄바꿈을 왜 한건가요 ?

잘못 누른거 같아요. 다시 지웠습니다!

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/CommentService.java:
```diff
@@ -52,16 +55,34 @@ private void validPageNumber(Integer pageNumber) {
+        if (writeCommentRequest.getParentId() == null) {
+            return;
+        }
+        if (commentRepository.findById(writeCommentRequest.getParentId()).get().getSolution().getId() != solutionId) {
```
왜 계속 검증없이 그냥 Optional값을 get으로 빼고 있는지 모르겠습니다.

jd99iam replied:
> 왜 계속 검증없이 그냥 Optional값을 get으로 빼고 있는지 모르겠습니다.

여기도 마찬가지로 이 메서드를 호출하기 전 라인에서 값이 있는지 검증합니다!

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/CommentService.java:
```diff
@@ -52,16 +55,34 @@ private void validPageNumber(Integer pageNumber) {
         Member member = findMemberById(memberId);
-        Comment comment = writeCommentRequest.toComment(member,solution);
+        validParentExists(writeCommentRequest);
+        validSameSolution(writeCommentRequest, solutionId);
```
둘 다 parendId가 null이면 검증하지 않는것 같은데 그걸 각각의 valid 메서드 내에서 할게 아니라 밖에서 한번 검증하고 각각 검증해야하지 않나요 ?
예를 들면 아래와 같습니다. 
```java
if ( writeCommentRequest.getParentId() != null) {
    validParentExists(writeCommentRequest);
    validSameSolution(writeCommentRequest, solutionId);
}
```

jd99iam replied:
> 둘 다 parendId가 null이면 검증하지 않는것 같은데 그걸 각각의 valid 메서드 내에서 할게 아니라 밖에서 한번 검증하고 각각 검증해야하지 않나요 ? 예를 들면 아래와 같습니다.
> 
> ```java
> if ( writeCommentRequest.getParentId() != null) {
>     validParentExists(writeCommentRequest);
>     validSameSolution(writeCommentRequest, solutionId);
> }
> ```

수정했습니다!!

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/ReviewService.java:
```diff
@@ -35,15 +39,46 @@ public class ReviewService {
         Solution solution = findSolutionById(solutionId);
-        reviewRepository.save(writeReviewRequest.toReview(writer,solution));
+        validParentExists(writeReviewRequest);
+        validSameSolution(writeReviewRequest,solutionId);
```
여기도 마찬가지로 parentId있는지 밖에서 검증하면 좋을 것 같아요

jd99iam replied:
> 여기도 마찬가지로 parentId있는지 밖에서 검증하면 좋을 것 같아요

수정했습니다!!

rlfrkdms1 reviewed (commented):
커밋 메세지가 너무 불친절한 것 같습니다 ㅜㅜ
조금 더 자세히 적어주면 좋을 것 같아요 ㅜㅜ
화이팅입니다.

jd99iam committed:
refactor: 라인 정리

한 라인에 반환값이 너무 많아서 라인 나눠줌

related to: #186

jd99iam committed:
refactor: 줄바꿈 해제

잘못 눌러서 생긴 줄바꿈 해제

related to: #186

jd99iam commented:
> 커밋 메세지가 너무 불친절한 것 같습니다 ㅜㅜ 조금 더 자세히 적어주면 좋을 것 같아요 ㅜㅜ 화이팅입니다.

네 더 자세히 적겠습니다!

jd99iam committed:
refactor: 검증 로직 리팩터링

검증 하는 여러 메서드에서 공통되게 부모 id가 null인지 체크하므로 해당 로직을 밖에서 공통되게 수행
- CommentService의 검증 메서드들 수정
- ReviewService의 검증 메서드들 수정

related to: #186
